jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/VirtualThreadsTest.java

```diff
@@ -71,4 +77,71 @@ public boolean handle(Request request, Response response, Callback callback)
 
         assertEquals(HttpStatus.OK_200, response.getStatus(), " for transport " + transport);
     }
+
+    @ParameterizedTest
+    @MethodSource("transports")
+    public void testBlockingClientListenersInvokedOnVirtualThread(Transport transport) throws Exception
+    {
+        testClientListeners(transport, true);
+    }
+
+    @ParameterizedTest
+    @MethodSource("transports")
+    public void testNonBlockingClientListenersInvokedOnPlatformThread(Transport transport) throws Exception
+    {
+        testClientListeners(transport, false);
+    }
+
+    private void testClientListeners(Transport transport, boolean blocking) throws Exception
+    {
+        startServer(transport, new Handler.Abstract()
+        {
+            @Override
+            public boolean handle(Request request, Response response, Callback callback) throws Exception
+            {
+                // Send only the headers.
+                response.write(false, null, Callback.NOOP);
+                // Wait to force the client to invoke the content
+                // callback separately from the headers callback.
+                Thread.sleep(500);
```

Comment:
No `sleep` please; it's better to wait until the headers event has been called.

Reply:
It is going to be really complicated.
We need to wait for the headers event, but guarantee that there is a `Content.Source.read()` that returns `null`.